File: app/src/main/java/com/medo/tweetspie/base/BaseFragment.java
package com.medo.tweetspie.base;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;

import com.medo.tweetspie.MainApp;
import com.medo.tweetspie.bus.BusProvider;
import com.medo.tweetspie.injection.components.AppComponent;


public abstract class BaseFragment extends Fragment {

  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {

    super.onCreate(savedInstanceState);
    inject(MainApp.getAppComponent());
  }

  @Override
  public void onResume() {

    super.onResume();
    // register the bus
    BusProvider.getInstance().register(this);
  }

  @Override
  public void onPause() {

    super.onPause();
    // unregister the bus
    BusProvider.getInstance().unregister(this);
  }

  protected abstract void inject(@NonNull AppComponent appComponent);
}

File: util/src/main/java/com/medo/tweetspie/util/base/BaseFragment.kt
package com.medo.tweetspie.util.base

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import com.medo.tweetspie.util.extensions.snack

abstract class BaseFragment : Fragment() {

    abstract val viewModel: BaseViewModel?

    abstract fun getLayoutId(): Int

    private val failureObserver = Observer<String?> { activity?.snack(it) }

    override fun onResume() {
        super.onResume()
        viewModel?.failure?.observe(this, failureObserver)
    }

    override fun onPause() {
        super.onPause()
        viewModel?.failure?.removeObserver(failureObserver)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? =
        inflater.inflate(getLayoutId(), container, false)
}